### DISCUSSION TITLE
New unofficial XState TypeScript decorators

### DISCUSSION THREAD

**Discussion by aitorllj93** [statelyai/xstate#2610] [General]
Hi! I just released a package on [npm](https://www.npmjs.com/package/@d3v0ps/xstate-decorators) with TypeScript decorators to be able to integrate XState seamlessly on dependency injection oriented frameworks like Angular or Nestjs. 
Right now, I have tested them with Nestjs and it works fine, you can inject dependencies on the class and use them in your guards definitions. I will test them in Angular too. I encourage you to try it and give me feedback in the form of issues, suggestions or contributions.

## Installation

```sh
npm install xstate reflect-metadata @d3v0ps/xstate-decorators
```

## Usage

```ts
import {
  StateMachine,
  Guard,
  StateMachineExecutor,
} from '@d3v0ps/xstate-decorators';

// defining the machine (machineOptions is optional)
@StateMachine<DoorMachineContext>(machineConfig, machineOptions)
class DoorMachineExecutor extends StateMachineExecutor<DoorMachineContext> {

  constructor(
    private checkIsAdmin: CheckIsAdmin
  ) {
    super();
  }

  @Guard('isAdmin')
  isAdmin(context: DoorMachineContext) {
    return this.checkIsAdmin.with(context);
  }

  @Guard('shouldAlert')
  shouldAlert(context: DoorMachineContext) {
    return context.alert === true;
  }
}
```

#### Running transitions

```ts
const doorMachineExecutor = new StateMachineExecutor(checkIsAdmin);

doorMachineExecutor.transition('closed', 'OPEN');
```

#### Using with outside context

i.e loading it from db or/and with Dependency Injection

```ts
const doorMachineExecutor = new StateMachineExecutor(checkIsAdmin);
const doorFromDb = {
  state: 'closed',
  context: {
    level: 'user',
    alert: false,
  },
};

doorMachineExecutor.transition(doorFromDb.state, 'OPEN', doorFromDb.context);
// this internally calls to
//    machine.withContext(context)
//           .withConfig(this.options)
//           .transition(currentState, event)
```

**Comment by davidkpiano** [comment]
This looks awesome, I'll try it out soon!

**Reply by aitorllj93** [reply]
Thank you! right now there are only guards decorators and transition method as my use case was very small, but I can add more features if needed

**Comment by VanTanev** [comment]
I am developing a project that uses exactly nestjs and xstate (FSM only, because that's what I needed for my use case).

I'm interested in sharing experiences, maybe over discord - I'm there with the same name.

The things I've found that aren't very obvious are, where to put async actions - it seems like doing invokes might be one possible way forward but it seems too convoluted. I opted to kind of do the opposite:

Define the machine as normal, and then define a class with all "entry points" into the machine. Then I do this:

```ts
const machine = createMachine<Context, Event>(
    {
        initial: 'initial',
        states: {
            [INITIAL]: {
                on: {
                    CREATE: [
                        {
                            cond: hasAssigneeSet,
                            target: IN_TAGGING,
                        },
                        {
                            target: READY_FOR_TAGGING,
                        },
                    ],
                },
            },
            [READY_FOR_TAGGING]: {
                on: {
                    SELF_ASSIGN: {
                        target: IN_TAGGING,
                    },
                },
            },
            [IN_TAGGING]: {},
        },
    },
    {},
)


class MachineService {
  async createWorkflow(assignee?: User) {
    let state = machine.transition('initial', { type: 'CREATE', assignee })
    if (!state.changed) throw new InvalidTransitionError()
    
    return WorkflowService.createFromState(state)
  }
  
  async selfAssign(workflow: Workflow) {
    let currentUser = await getCurrentUser()
    let state = machine.transition(workflow.state, { type: 'SELF_ASSIGN', currentUser })
    if (!state.changed) throw new InvalidTransitionError()
    
    return WorkflowService.persistTransition(workflow, state.value)
  }
}
```

Basically, I just do control flow, but I do actual execution outside of the state machine. Thoughts?

**Reply by aitorllj93** [reply]
Hi, thank you for your feedback, I have looked onto the `invoke` syntax and it's something I would like to implement, but as you said it's quite convoluted. The main issue from my side is that, as it's inside the machine json, I would need to iterate over it to assign the class method call to the corresponding state, I wish there was another way... I will implement `actions` soon as they are the same than guards in terms of implementing them.

About your provided example, with the current state of the library it could be like this:

```ts

import { StateMachine, Guard, StateMachineExecutor } from '@d3v0ps/xstate-decorators';

@StateMachine<Context>({
  initial: 'initial',
  states: {
    [INITIAL]: {
      on: {
        CREATE: [
          {
            cond: 'hasAssigneeSet',
            target: IN_TAGGING,
          },
          {
            target: READY_FOR_TAGGING,
          },
        ],
      },
    },
    [READY_FOR_TAGGING]: {
      on: {
        SELF_ASSIGN: {
          target: IN_TAGGING,
        },
      },
    },
    [IN_TAGGING]: {},
  },
})
export class MachineService extends StateMachineExecutor<Context> {

  constructor(private workflowService: WorkflowService) {
    super();
  }

  @Guard('hasAssigneeSet')
  hasAssigneeSet(context: Context) {
    return context.assignee ? true : false;
  }


  async createWorkflow(assignee?: User) {
    let state = this.transition('initial', { type: 'CREATE', assignee });
    // You can use this.machine.transition to use the original xstate method instead
    //
    // Remember that if you don't pass a third parameter with the context, the context will be reused everytime you call the machine as it's the same instance
    // let state = this.transition('initial', { type: 'CREATE', assignee }, {}); // you can pass an empty context
    // You can also create new instances of the machine by calling the `with` method
    // const machine = this.with(context) or this.machine.withContext(context).withConfig(this.config)

    if (!state.changed) throw new InvalidTransitionError();
    
    return this.workflowService.createFromState(state);
  }
  
  async selfAssign(workflow: Workflow) {
    let currentUser = await getCurrentUser();
    let state = this.transition(workflow.state, { type: 'SELF_ASSIGN', currentUser });
    if (!state.changed) throw new InvalidTransitionError();
    
    return this.workflowService.persistTransition(workflow, state.value)
  }
}
```

**Reply by aitorllj93** [reply]
I just saw I can pass services in the configuration and reference them with strings on the invoke src property, so I will add `invoke` support soon!

**Comment by aitorllj93** [comment]
Hi, I just released `Interpreter` decorator and `Action`, `Activity` and `Service` decorators too.

```ts

import {
  Action,
  Activity,
  Interpreter,
  Guard,
  Service,
  StateMachineInterpreter,
} from '@d3v0ps/xstate-decorators';

...

@Interpreter<DoorMachineContext>({
  ...
})
export class DoorMachineInterpreter extends StateMachineInterpreter<DoorMachineContext> {

  @Action('logEvent')
  logEvent(context: DoorMachineContext, event) {
    console.log(event);
  }

  @Activity('sayHello')
  sayHello() {
    this.helloService.sayHello();
  }

  @Guard('isAdmin')
  isAdmin(context: DoorMachineContext) {
    return this.checkIsAdmin.with(context);
  }

  @Service('getDoor')
  async getDoor(context: DoorMachineContext): Promise<DoorMachineContext> {
    return this.doorService.getDoor(context);
  }
}

```
